core/src/storage/impls/state.rs

```diff
@@ -276,9 +285,124 @@ impl<'a> StateTrait for State<'a> {
+        let mut upper_bound_excl_value = lower_bound_incl.clone();
+        let upper_bound_excl = if lower_bound_incl.len() == 0 {
+            None
+        } else {
+            let mut carry = 1;
+            let len = upper_bound_excl_value.len();
+            for i in 0..len {
+                if upper_bound_excl_value[len - 1 - i] == 255 {
+                    upper_bound_excl_value[len - 1 - i] = 0;
+                } else {
+                    upper_bound_excl_value[len - 1 - i] += 1;
+                    carry = 0;
+                    break;
+                }
+            }
+            // all bytes in lower_bound_incl are 255, which means no upper bound
+            // is needed.
+            if carry == 1 {
+                None
+            } else {
+                Some(upper_bound_excl_value.as_slice())
             }
+        };
+        let mut kvs = kv_iterator
+            .iter_range(lower_bound_incl.as_slice(), upper_bound_excl)?;
+
+        let mut snapshot_kvs = Vec::new();
+        while let Some((key, value)) = kvs.next()? {
+            snapshot_kvs.push((key, value.into_boxed_slice()));
         }
+
+        let mut result = Vec::new();
+        // This is used to keep track of the deleted keys.
+        let mut deleted_keys = HashSet::new();
+        if let Some(kvs) = delta_trie_kvs {
+            for (k, v) in kvs {
+                deleted_keys.insert(k.clone());
+                if v.len() > 0 {
+                    result.push((k, v));
+                }
+            }
+        }
+
+        if let Some(kvs) = intermediate_trie_kvs {
+            for (k, v) in kvs {
+                let storage_key = StorageKey::from_key_bytes(&k);
+                // Only delete nonempty keys.
+                if v.len() > 0 {
+                    self.delete(storage_key)?;
+                }
+                // Convert the key to `delta_mpt_key_bytes` used in Delta Trie
+                // for consistency.
+                let k = storage_key
+                    .to_delta_mpt_key_bytes(&self.delta_trie_key_padding);
+                if !deleted_keys.contains(&k) {
+                    deleted_keys.insert(k.clone());
+                    if v.len() > 0 {
+                        result.push((k, v));
+                    }
+                }
+            }
+        }
+
+        for (k, v) in snapshot_kvs {
+            let storage_key = StorageKey::from_key_bytes(&k);
+            // Only delete nonempty keys.
+            if v.len() > 0 {
+                self.delete(storage_key)?;
+            }
+            // Convert the key to `delta_mpt_key_bytes` used in Delta Trie
+            // for consistency.
+            let k = storage_key
+                .to_delta_mpt_key_bytes(&self.delta_trie_key_padding);
+            if !deleted_keys.contains(&k) {
+                if v.len() > 0 {
+                    result.push((k, v));
+                }
+            }
+        }
+
+        Ok(Some(result))
```

Comment:
Let's use the snapshot_key (the actual key) for the result?

It's correct to convert to delta_mpt_key for map lookup but it wastes computation resources.